java/vector/src/main/java/org/apache/arrow/vector/complex/impl/PromotableWriter.java

```diff
@@ -118,21 +126,22 @@ public void setPosition(int index) {
     }
   }
 
-  protected FieldWriter getWriter(MinorType type) {
+  @Override
+  protected FieldWriter getWriter(ArrowType type) {
     if (state == State.UNION) {
-      ((UnionWriter)writer).getWriter(type);
+      ((UnionWriter) writer).getWriter(Types.getMinorTypeForArrowType(type));
     } else if (state == State.UNTYPED) {
       if (type == null) {
         // ???
         return null;
       }
-      ValueVector v = listVector.addOrGetVector(FieldType.nullable(type.getType())).getVector();
+      ValueVector v = promotableVector.addOrGetVector(FieldType.nullable(type)).getVector();
       v.allocateNew();
       setWriter(v);
       writer.setPosition(position);
-    } else if (type != this.type) {
+    } else if (!Objects.equals(type, this.type)) {
       promoteToUnion();
-      ((UnionWriter)writer).getWriter(type);
+      ((UnionWriter) writer).getWriter(Types.getMinorTypeForArrowType(type));
```

Comment:
There's a discrepancy here. We compare ArrowType but then we create the writer based on MinorType which the two types may have the same even if they are not equal (Decimal or FixedWidthList for example).
It looks like if we start writing fixed width lists and then we write a different width we will promote but write the different width lists to the old fixed width vector. Possibly we should fail in this case.